packages/flutter/lib/src/widgets/app.dart

```diff
@@ -690,35 +734,144 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
 
   // LOCALIZATION
 
+  /// This is the resolved locale, and is one of the supportedLocales.
   Locale _locale;
 
-  Locale _resolveLocale(Locale newLocale, Iterable<Locale> supportedLocales) {
+  Locale _resolveLocales(List<Locale> preferredLocales, Iterable<Locale> supportedLocales) {
+    // Attempt to use localeListResolutionCallback.
+    if (widget.localeListResolutionCallback != null) {
+      final Locale locale = widget.localeListResolutionCallback(preferredLocales, widget.supportedLocales);
+      if (locale != null)
+        return locale;
+    }
+    // localeListResolutionCallback failed, falling back to localeResolutionCallback.
     if (widget.localeResolutionCallback != null) {
-      final Locale locale = widget.localeResolutionCallback(newLocale, widget.supportedLocales);
+      final Locale locale = widget.localeResolutionCallback(preferredLocales.first, widget.supportedLocales);
       if (locale != null)
         return locale;
     }
-    // newLocale can be null when called before the platform has had a chance to
-    // initialize the locales. We default to the first supported locale.
-    if (newLocale == null) {
+    // Both callbacks failed, falling back to default algorithm.
+    return fallbackLocaleResolution(preferredLocales, supportedLocales);
+  }
+
+  /// The default fallback locale resolution algorithm.
+  ///
+  /// This algorithm prioritizes speed at the cost of slightly less appropriate
+  /// resolutions for edge cases.
+  ///
+  /// Custom resolution algorithms can be provided through [WidgetsApp.localeListResolutionCallback]
+  /// or [WidgetsApp.localeResolutionCallback].
+  ///
+  /// This algorithm will resolve to the earliest locale in [preferredLocales] that
+  /// matches the most fields, prioritizing in the order of perfect match,
+  /// languageCode+countryCode, languageCode+scriptCode, languageCode-only.
+  ///
+  /// In the case where a locale is matched by languageCode-only and is not the
+  /// default (first) locale, the next locale in preferredLocales with a
+  /// perfect match can supercede the languageCode-only match if it exists.
+  ///
+  /// When a preferredLocale matches more than one supported locale, it will resolve
+  /// to the first matching locale listed in the supportedLocales.
+  ///
+  /// When all [preferredLocales] have been exhausted without a match, the first countryCode only
+  /// match will be returned.
+  ///
+  /// When no match at all is found, the first (default) locale in [supportedLocales] will be
+  /// returned.
+  ///
+  /// This algorithm does not take language distance (how similar languages are to each other)
+  /// into account, and will not handle edge cases such as resolving `de` to `fr` rather than `zh`
+  /// when `de` is not supported and `zh` is listed before `fr` (German is closer to French
+  /// than Chinese).
+  static Locale fallbackLocaleResolution(List<Locale> preferredLocales, Iterable<Locale> supportedLocales) {
```

Comment:
I agree that this is not quite a "fallback" but "selectPreferredSupportedLocale" seems to give off the impression that our algorithm is superior because it is "preferred". It also may cause confusion where they think our algo is "preferred" over the user-provided callbacks.

I would feel better about a name along the lines of `defaultLocaleListResolution()` to maintain consistency with the naming of `localeListResolutionCallback` and `localeResolutionCallback`